src/engine/guidance/post_processing.cpp

```diff
@@ -319,41 +328,63 @@ void trimShortSegments(std::vector<RouteStep> &steps, LegGeometry &geometry)
     // If a route from b to c is requested, both a--b and b--c could be selected as start segment.
     // In case of a--b, we end up with an unwanted turn saying turn-right onto b-c.
     // These cases start off with an initial segment which is of zero length.
-    // We have to be careful though, since routing that starts in a roundabout has a valid
-    // initial segment of length zero and we cannot delete the upcoming segment.
+    // We have to be careful though, since routing that starts in a roundabout has a valid.
+    // To catch these cases correctly, we have to perform trimming prior to the post-processing
 
-    if (steps.front().distance <= std::numeric_limits<double>::epsilon() &&
-        !entersRoundabout((steps.begin() + 1)->maneuver.instruction))
+    BOOST_ASSERT(geometry.locations.size() >= steps.size());
+    const bool zero_length_step = steps.front().distance <= std::numeric_limits<double>::epsilon();
+    const bool duplicated_coordinate = geometry.locations.at(0) == geometry.locations.at(1);
```

Comment:
We assert on `steps.size() >= 2` and `geometry.locations.size() >= steps.size()`, thus `geometry.locations.size() >= 2`, therefore you can change `.at(x)` to `[x]` for x in 0, 1.